enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+*Bootstrap phase:*
+
+- Initial configuration with static pods is expected to be using `ConfigMap` credentials, similarly to what is [done](https://github.com/openshift/installer/blob/ba6d7fe087eada5a4fc260064c85a484a8c45aaf/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L162) in `kcm` operator.
+- Once it is supported, the static pods will also use the `CredentialsRequest` to get cloud credentials.
+
+*Post-install phase:*
+
+- Using `CredentialsRequest` is the default option, every supported `CCM` within the `Deployment` will request its own set of credentials.
+- A set of `CredentialsRequest` resources will be hosted under `CCCMO` repository, and created by the `CVO`, similarly to the [machine-api](https://github.com/openshift/machine-api-operator/blob/6f629682b791a6f4992b78218bfc6e41a32abbe9/install/0000_30_machine-api-operator_00_credentials-request.yaml) approach.
+
+*Alternative approach (long-term)*
+
+An upstream effort in this [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md) attempts to remove `Kubelet` dependency on in-tree credentials issuing process, and instead use a similar approach with pluggable credentials providers. This interface is supposed to be moved into `Alpha` stage with this [PR](https://github.com/kubernetes/kubernetes/pull/94196) in `1.20` release.
+
+### Upgrade/Downgrade strategy
+
+A set of `cloud-controller-manager` pods will be running under a `Deployment` resource, provisioned by the operator. The leader-election will preserve the leadership during the cluster updates over cloud-specific loops.
+
+Upgrade from previous versions of OpenShift on OpenStack will look like:
+
+#### 4.6 -> 4.7
+
+Upgrade (all platforms):
+
+- The `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
+
+Downgrade (all platforms):
+
+- `cluster-cloud-controller-manager-operator` de-provisions CCM `deployments` and stops working.
+
+The steps below apply to OpenStack only:
+
+Upgrade:
+
+- The `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
+- `cluster-cloud-controller-manager-operator` creates all required resources for CCM (RBAC, Service Account, etc.).
+- `kube-apiserver` and `kube-controller-manager` are restarted without the `--cloud-provider` option.
+- `kubelet` is restarted with `--cloud-provider=external` option.
+
+Downgrade:
+
+- `cluster-cloud-controller-manager-operator` de-provisions CCM `deployments` and stops working.
+- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider=<some-platform>` option.
+- An old machine configuration is applied, which causes `kubelet` to restart with `--cloud-provider=<some-platform>` option.
+
+#### 4.7 -> 4.8
+
+Upgrade:
+
+- `cluster-cloud-controller-manager-operator` creates resources for any in-tree provider defaulting to this in 4.8 (expecting: AWS, Azure, GCP, vSphere).
```

Comment:
How this is different than the previous one? I'd like to see a timeline (can be a proposal) when each provider will be migrated to out-of-tree. Example:
- 4.7: GCP
- 4.8: AWS
- 4.9: all
etc.

Reply:
The release for each provider is going to depend on CSI plugin and migration readiness. Once both are shipped with the release, out-of-tree implementation will follow. The upper boundary is the exclusion of the in-tree upstream (starting in 1.23 for kubelet as of now) - appr 4.10.